src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -134,12 +134,33 @@ private bool IsIrrefutablePatternTest(BoundExpression test)
 
         private BoundExpression CompareWithConstant(BoundExpression input, BoundExpression boundConstant)
         {
-            return _factory.StaticCall(
-                _factory.SpecialType(SpecialType.System_Object),
-                "Equals",
-                _factory.Convert(_factory.SpecialType(SpecialType.System_Object), boundConstant),
-                _factory.Convert(_factory.SpecialType(SpecialType.System_Object), input)
-                );
+            var systemObject = _factory.SpecialType(SpecialType.System_Object);
+            if (boundConstant is BoundLiteral bl && bl.ConstantValue?.IsNull == true)
```

Comment:
The BoundLiteral test is incorrect. This should only depend on whether the constant value is null (whether or not it is a literal). For example, the right-hand-side could be a named constant. And we should have a test for that scenario (with type `string`).